src/modules/Game.class.js

```diff
@@ -1,68 +1,152 @@
-   *  [0, 0, 0, 0]]
-   *
-   * If passed, the board will be initialized with the provided
-   * initial state.
-   */
-  constructor(initialState) {
-    // eslint-disable-next-line no-console
-    console.log(initialState);
+  static Status = {
+    idle: 'idle',
+    playing: 'playing',
+    win: 'win',
+    lose: 'lose',
+  };
+
+  constructor(initialState = INITIAL_STATE) {
+    this.state = initialState.map((row) => [...row]);
+    this.status = Game.Status.idle;
+    this.initialState = initialState;
+    this.score = 0;
+  }
+
+  filterZero(row) {
+    return row.filter((num) => num !== 0);
+  }
+  slide(row, direction) {
+    let updRow = this.filterZero(row);
+
+    for (let i = 0; i < updRow.length - 1; i++) {
+      if (updRow[i] === updRow[i + 1]) {
+        updRow[i] *= 2;
+        updRow[i + 1] = 0;
+        this.score += updRow[i];
+      }
+    }
+    updRow = this.filterZero(updRow);
+
+    while (updRow.length < 4) {
+      updRow.push(0);
+    }
+
+    return ROW_MODIFIER_FUNC[direction](updRow);
   }
 
-  moveLeft() {}
-  moveRight() {}
-  moveUp() {}
-  moveDown() {}
-
-  /**
-   * @returns {number}
-   */
-  getScore() {}
-
-  /**
-   * @returns {number[][]}
-   */
-  getState() {}
-
-  /**
-   * Returns the current game status.
-   *
-   * @returns {string} One of: 'idle', 'playing', 'win', 'lose'
-   *
-   * `idle` - the game has not started yet (the initial state);
-   * `playing` - the game is in progress;
-   * `win` - the game is won;
-   * `lose` - the game is lost
-   */
-  getStatus() {}
-
-  /**
-   * Starts the game.
-   */
-  start() {}
-
-  /**
-   * Resets the game.
-   */
-  restart() {}
-
-  // Add your own methods here
+  move(direction) {
+    const rowMethod = ROW_MODIFIER_FUNC[direction];
+    const matrixMethod = MATRIX_MODIFIER_FUNC[direction];
+
+    if (this.status === Game.Status.playing) {
+      const updatedCells = matrixMethod(this.state).map((row) => {
+        return this.slide(rowMethod(row), direction);
+      });
+
+      this.state = matrixMethod(updatedCells);
+
+      this.getRandomCell();
+      this.checkGameStatus();
+    }
+  }
+
+  getScore() {
+    return this.score;
+  }
+  getState() {
+    return this.state;
+  }
+  getStatus() {
+    return this.status;
+  }
+
+  start() {
+    this.status = Game.Status.playing;
+    this.getRandomCell();
+    this.getRandomCell();
+  }
+
+  restart() {
+    this.status = Game.Status.idle;
+    this.state = this.initialState.map((row) => [...row]);
+    this.score = 0;
+  }
+
+  getRandomCell() {
+    const emptyCells = [];
+
+    for (let r = 0; r < 4; r++) {
+      for (let c = 0; c < 4; c++) {
+        if (this.state[r][c] === 0) {
+          emptyCells.push([r, c]);
+        }
+      }
+    }
+
+    if (emptyCells.length > 0) {
+      const [randomR, randomC] =
+        emptyCells[Math.floor(Math.random() * emptyCells.length)];
+
+      this.state[randomR][randomC] = Math.random() < 0.9 ? 2 : 4;
+    }
+  }
+  checkGameStatus() {
```

Comment:
Maybe, and this is a huge MAYBE, which means not mandatory,  you could dismantle this function and keep tracking of the game status during other functions, so you could remove some iterations in the table.
How? `hasEmptyCells` can be checked whenever a row gets iterated in, let's say, the `slide` method.
`isMovable` can also be checked during `slide` and `transpose`